### PR TITLE
[TBBAS-1170] IInstance update: support argument struct in constructor

### PR DESCRIPTION
**Pull request by denise-opendaq**
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description:

### CODE REVIEW

File: core/opendaq/opendaq/src/instance_builder_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    if (!this->moduleManager.assigned())
+        this->moduleManager = ModuleManager(getModuleManagerOptions()["ModulesPath"]);
+
+    const auto builderPtr = this->borrowPtr<InstanceBuilderPtr>();
+    return daqTry([&]()
+    {
+        *instance = InstanceFromBuilder(builderPtr).detach();
+        return OPENDAQ_SUCCESS;
+    });
+}
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/src/instance_builder_impl.cpp:96]: The entire building logic should be in the Instance constructor. This would allow a custom implementation of a builder to be provided. The `build` method should only call the appropriate factory.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    /*!
+     * @brief Sets the Logger level of Instance component
+     * @param component The name of Instance component
+     * @param logLevel The log level of Instance component
+     */
+    virtual ErrCode INTERFACE_FUNC setComponentLogLevel(IString* component, LogLevel logLevel) = 0;
+
+    // [returnSelf]
+    /*!
+     * @brief Sets the sink logger level of default Instance logger. Ignored if was set custom logger
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:77]: Grammar in comment -> "Ignored if a custom logger has already been set". Similar in other comments.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+     * @param[out] option The dictionary of options of Instance
+     */
+    virtual ErrCode INTERFACE_FUNC getOptions(IDict** options) = 0;
+
+    // [returnSelf]
+    /*!
+     * @brief Sets the root device of Instance
+     * @param rootDevice The root device of Instance
+     */
+    virtual ErrCode INTERFACE_FUNC setRootDevice(IDevice* rootDevice) = 0;
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:143]: These should accept an IString parameter, indicating the connection string.

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]
Good start to the task, and the functionality is there. We should do a review of how the builder and each separate method is planned on being used, and do a small re-design of the builder for it to be tailored towards end-users.

Also, I think this PR requires a How-to guide explaining how to use the builder, and what each specific builder field does, as well as what the dependencies between different fields are.

File: core/opendaq/opendaq/include/opendaq/instance_factory.h [hunk 4]
```diff
@@ -75,9 +96,9 @@ inline InstancePtr Instance(const std::string& modulePath = "", const std::strin
  * The Client supports adding function blocks and connecting to devices. The array of function blocks
  * and supported devices depends on the modules loaded by the Module manager.
  */
-inline DevicePtr Client(const ContextPtr& context, const StringPtr& localId)
+inline DevicePtr Client(const ContextPtr& context, const StringPtr& localId, const DeviceInfoPtr& defaultDeviceInfo = DeviceInfoPtr())
```
[the remark this answers is not in the document]
Comment (denise-opendaq) [reply core/opendaq/opendaq/include/opendaq/instance_factory.h:99]: DeviceInfoPtr() will create empty pointer

File: core/opendaq/opendaq/include/opendaq/instance.h [hunk 5]
```diff
@@ -22,6 +22,8 @@
 
 BEGIN_NAMESPACE_OPENDAQ
 
+struct IInstanceBuilder;
+
 /*!
  * @ingroup opendaq_devices
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance.h:28]: Should we remove the `setRootDevice` call from IInstance?

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]

File: core/opendaq/opendaq/include/opendaq/instance.h (hunk 5 above)
Comment (dejancrn) [reply core/opendaq/opendaq/include/opendaq/instance.h:28]: I think we can do that

File: core/opendaq/opendaq/src/instance_builder_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    this->logger = logger;    
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode InstanceBuilderImpl::getLogger(ILogger** logger)
+{
+    if (logger == nullptr)
+        return OPENDAQ_ERR_ARGUMENT_NULL;
+    
+    if (this->logger.assigned())
```
Comment (dejancrn) [inline core/opendaq/opendaq/src/instance_builder_impl.cpp:79]: logger.addRefAndReturn() will return nullptr if logger not assigned. No need to handle not assigned case here.

File: core/opendaq/opendaq/src/instance_builder_impl.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        return OPENDAQ_ERR_ARGUMENT_NULL;
+    
+    if (this->scheduler.assigned())
+        *scheduler = this->scheduler.addRefAndReturn();
+    else
+        *scheduler = nullptr;
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode InstanceBuilderImpl::setDefaultRootDeviceName(IString* localId)
```
Comment (dejancrn) [inline core/opendaq/opendaq/src/instance_builder_impl.cpp:227]: function should be called set/getDefaultRootDeviceLocalId

**Review by dejancrn — APPROVED** [review]
At some point doc comments should be checked with some grammar and spell checker

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+     * @brief Gets the default device info of Instance
+     * @param defaultDevice The default device info of Instance
+     */
+    virtual ErrCode INTERFACE_FUNC getDefaultRootDeviceInfo(IDeviceInfo** deviceInfo) = 0;
+};
+/*!@}*/
+
+/*!
+ * @brief Creates a InstanceBuilder with no parameters configured.
+ */
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:246]: This comment should be in a Doxygen group

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+ * object before its creation. It allows customization of the logger, module manager, scheduler and root device. 
+ * Once configured, calling the `build` method returns a fully initialized Instance object with the specified settings.
+ * 
+ * @subsection Configuration Methods:
+ * The InstanceBuilder provides the following configuration methods:
+ *
+ * - **Logger:** The custom Logger for the Instance. This logger will be used for logging messages related to
+ *   the Instance and its components. When configured, the GlobalLogLevel, ComponentLogLevel and LoggerSink will be ignored, as they are in use only with the default Instance logger
+ *
+ * - **GlobalLogLevel:** The default Logger global log level for the Instance. All log messages with a severity
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:43]: I'd lean towards writing the names of properties/builder fields with spaces in-between and only the 1st word capitalized. Eg. **Global log level**. This is how the Property Doxygen documentation was structured.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+ * - **ModulePath:** The path for the default ModuleManager of the Instance. If a custom module manager has not
+ *   been set, this path will be used to load modules.
+ *
+ * - **Scheduler:** The custom scheduler for the Instance. If set, the number of worker threads will be ignored.
+ *
+ * - **SchedulerWorkerNum:** The number of worker threads in the scheduler of the Instance. If a scheduler has
+ *   not already been set, this defines the number of threads available for parallel processing.
+ *
+ * - **DefaultRootDeviceName:** The virtual Client as the default root device with the specified name. If the
+ *   RootDevice has not been set, the Instance Root Device will also be set as the virtual Client.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:64]: The wording here is a bit unclear. What is the "virtual Client"? This should clearly state that the name of the default client root device is modified when this is set.

Also, should this not be "Default root device local ID"?

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+ *
+ * - **DefaultRootDeviceInfo:** The default device information for the Instance. If the device information has
+ *   not been set, the `getInfo` method of the Instance will return this set device information.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:67]: This does not modify the device info of the instance, but of the default root device.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+ *
+ * - **RootDevice:** The custom root device of the Instance from the connection string. This defines the device that
+ *   serves as the entry point for the Instance.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:70]: When this is set, DefaultRootDeviceInfo and DefaultRootDeviceName are ignored.

I'd word this as:
"The connection string of a device that replaces the default openDAQ root device (virtual client). When the instance is created a connection to the device with the given connection string will be established, and the device will be placed at the root of the component tree structure."

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+ *
+ * - **SinkLogLevel:** The sink logger level of the default Instance logger. This level is ignored if a custom
+ *   logger has already been set.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:73]: This should probably be moved next to the other logger documentation entries.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+ *
+ * Note:
+ * Some configuration options, such as the custom logger, module manager, and scheduler, are optional. If not set, the
+ * Instance will use default implementations. Users can choose to configure only the specific aspects they need,
+ * allowing for a tailored setup.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:78]: Aren't all of the options optional?

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+ */
+DECLARE_OPENDAQ_INTERFACE(IInstanceBuilder, IBaseObject)
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:80]: The same comments as above apply for the setters/getters in the interface.

File: core/opendaq/opendaq/src/instance_builder_impl.cpp (hunk 6 above)
Comment (JakaMohorkoDS) [reply core/opendaq/opendaq/src/instance_builder_impl.cpp:79]: This still applies - the `if (this->logger.assigned())` check is redundant here

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+ * The InstanceBuilder provides the following configuration methods:
+ *
+ * - **Logger:** The custom Logger for the Instance. This logger will be used for logging messages related to
+ *   the Instance and its components. When configured, the GlobalLogLevel, ComponentLogLevel and LoggerSink will be ignored, as they are in use only with the default Instance logger
+ *
+ * - **GlobalLogLevel:** The default Logger global log level for the Instance. All log messages with a severity
+ *   level equal to or higher than the specified level will be processed.
+ *
+ * - **ComponentLogLevel:** The Logger level for a specific component of the Instance. Log messages related to
+ *   that component will be processed according to the specified log level.
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:47]: Log messages related to that component ~~will be processed according to the specified log level.~~ with a severity level equal or higher than the specified level will be processed.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+/*!
+ * @brief Builder component of Instance objects. Contains setter methods to configure the Instance parameters, such as Context (Logger, Scheduler, ModuleManager) and RootDevice.
+ * Contains a  `build` method that builds the Instance object.
+ *
+ * The InstanceBuilder provides a fluent interface for setting various configuration options for an Instance
+ * object before its creation. It allows customization of the logger, module manager, scheduler and root device. 
+ * Once configured, calling the `build` method returns a fully initialized Instance object with the specified settings.
+ * 
+ * @subsection Configuration Methods:
+ * The InstanceBuilder provides the following configuration methods:
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/include/opendaq/instance_builder.h:38]: I'm really missing the defaults of each of these values. What happens when none of the fields are set? Eg. how many worker threads does the scheduler use if not configured?

File: core/opendaq/opendaq/src/instance_impl.cpp [hunk 18]
```diff
@@ -26,6 +26,62 @@ InstanceImpl::InstanceImpl(ContextPtr context, const StringPtr& localId)
...
+    const auto builderPtr = InstanceBuilderPtr::Borrow(instanceBuilder);
+    
+    auto localId = builderPtr.getDefaultRootDeviceLocalId();
+    auto instanceId = defineLocalId(localId.assigned() ? localId.toStdString() : std::string());
+    defaultRootDevice = Client(this->context, instanceId, builderPtr.getDefaultRootDeviceInfo());
+    rootDevice = defaultRootDevice;
+
+    auto connectionString = builderPtr.getRootDevice();
+    if (connectionString.assigned())
+        rootDevice = detail::createDevice(connectionString, nullptr, nullptr, moduleManager, nullptr);
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/src/instance_impl.cpp:80]: Nitpicking, but the `defaultRootDevice` should probably not be assigned to `rootDevice` at all if the `connectionString` is assigned. I'd also add another check for the `connectionString` being empty and ignore it in that case.

File: core/opendaq/opendaq/tests/test_instance.cpp [hunk 19]
```diff
@@ -307,5 +307,109 @@ TEST_F(InstanceTest, Serialize)
     std::cout << str << std::endl;
 }
 
+TEST_F(InstanceTest, InstanceBuilderSetGet)
```
Comment (JakaMohorkoDS) [inline core/opendaq/opendaq/tests/test_instance.cpp:310]: I'd just add a test that validates that the correct the correct fields are ignored when setting incompatible options. Eg. that the "Scheduler worker num" is ignored when the Scheduler is set.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 20]
```diff
@@ -0,0 +1,291 @@
+= Instance configuration
+
+The Instance is the top-level openDAQ(TM) object, serving as a container for the openDAQ context and the base module manager.
+
+**Related articles:**
+
+  * xref:background_info:opendaq_architecture.adoc#instance[Architecture - Instance]
+
+== Creating the default openDAQ(TM) Instance
+Upon creation, it generates a Client device, which is a default device implementation capable of loading any function blocks present in the module manager's search path. If the native openDAQ client module is loaded, the Client device can connect to any TMS-enabled device using the `addDevice` function. The Client is set as the root device when the instance is created.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:10]: I'd modify this a bit. The client device is capable of using the module manager to add function blocks, and connect to any device supported by the loaded modules. This is not OPC UA specific, but just depends on the modules loaded. Eg. you can connect to any streaming-only device that will not use OPC UA. 

Notably, "TMS" is a deprecated keyword, the OPC UA standard is now named the "openDAQ OPC UA" standard.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+----
+#include <opendaq/opendaq.h>
+#include <iostream>
+
+int main(int argc, const char* argv[])
+{
+    daq::InstancePtr instance = daq::Instance();
+    daq::DevicePtr device = instance.addDevice("daq.opcua://127.0.0.1/");
+
+    return 0;
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:25]: This code sample is probably redundant for the scope of this guide - it should more be focused on the builder.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+[source,cpp]
+----
+#include <opendaq/opendaq.h>
+#include <iostream>
+
+int main(int argc, const char* argv[])
+{
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder()
+    daq::InstancePtr instance = instanceBuilder.build();
+    daq::DevicePtr device = instance.addDevice("daq.opcua://127.0.0.1/");
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:46]: Again, I'd leave out the `addDevice` and other instance calls here. I'd explain what happens when you call `build`. In this case I'd mention that the defaults are used, and what the defaults are.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc (hunk 20 above)
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:10]: This is missing information on what the defaults are.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+#include <iostream>
+
+int main(int argc, const char* argv[])
+{
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder();
+    instanceBuilder.setGlobalLogLevel(daq::LogLevel::Warn);
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    return 0;
+}
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:72]: I'd ommit the wrapper `main` function call, and just leave the important lines of code:

```
    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder();
    instanceBuilder.setGlobalLogLevel(daq::LogLevel::Warn);
    daq::InstancePtr instance = instanceBuilder.build();
```

In general, I'd do that in all but the first code sample.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+
+**Related articles:**
+
+  * xref:background_info:opendaq_architecture.adoc#instance[Architecture - Instance]
+
+== Creating the default openDAQ(TM) Instance
+Upon creation, it generates a Client device, which is a default device implementation capable of loading any function blocks present in the module manager's search path. If the native openDAQ client module is loaded, the Client device can connect to any TMS-enabled device using the `addDevice` function. The Client is set as the root device when the instance is created.
+[tabs]
+====
+Cpp::
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:13]: Code examples should be available in Python as well.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+                        .setComponentLogLevel("Some Module", daq::LogLevel::Off);
+
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    return 0;
+}
+----
+====
+
+An advanced way of configuring the Instance logger is to configure the logger object and set it with the `setLogger` method.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:101]: I'd try to avoid the mention of language-specific function call notation, as it becomes harder to write documentation for other languages this way. I'd reference function calls only in  code samples.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+----
+====
+
+[NOTE]
+====
+Alternatively, a developer can create their own implementation of the logger, which has to inherit the ILogger interface.
+====
+
+== Configure openDAQ(TM) Instance Module Manager
+By default, the Instance Module Manager uses the current path for loading modules. In the Instance Builder, developers can set a custom module path or replace the default Module Manager with a custom implementation inherited from IModuleManager.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:131]: The "inherited from IModuleManager" part of the sentence should be left out. If left in, it should read as "with a custom implementation that implements the IModuleManager interface".

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+{
+    daq::ModuleManagerPtr moduleManager = daq::ModuleManagerPtr(CustomMuduleManagerImpl());
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder().setModuleManager(moduleManager);
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    return 0;
+}
+----
+Setting custom module manager
+====
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:178]: This code is out-of-scope imho. I'd just leave it out, as custom implementations of interfaces are not yet documented at all and are not the target of this how-to.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+int main(int argc, const char* argv[])
+{
+    daq::SchedulerPtr scheduler = daq::SchedulerPtr(CustomSchedulerImpl(2));
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder().setScheduler(scheduler);
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    return 0;
+}
+----
+====
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:228]: Same as above.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+
+== Configure openDAQ(TM) Root Device
+If, after creating the Instance object, the developer has not set the root device, the Instance will use the Client device as the root. Using the Instance Builder, developers can discover and connect to available devices using the method `getAvailableDevices`. In the Instance Builder, developers can forcibly set the available device.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:231]: These two sentences feel a bit off. The Client device is the default root that queries the module manager when adding/searching for function blocks/devices. The builder allows a user to specify a different root device that replaces the default, or to change the local ID and info of the default.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 30]
```diff
@@ -0,0 +1,291 @@
...
+
+int main(int argc, const char* argv[])
+{
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder().setSchedulerWorkerNum(2);
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    return 0;
+}
+----
+====
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:249]: This code example seems unrelated to the current documentation section?

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 31]
```diff
@@ -0,0 +1,291 @@
...
+As in the previous examples, developers can implement their own scheduler solution and use it in Instance Builder. In that mode, developers can discover and connect to an available device using method `getAvailableDevices`. In Instance Builder, developers can force set available devices.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:250]: Similar issue with the sentence here as described above.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 32]
```diff
@@ -0,0 +1,291 @@
...
+#include <opendaq/opendaq.h>
+#include <iostream>
+
+int main(int argc, const char* argv[])
+{
+    daq::DeviceInfoPtr defaultRootDeviceInfo = daq::DeviceInfo("daqref://device0");
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder().setDefaultRootDeviceInfo(defaultRootDeviceInfo);
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    assert(instance.getAvailableDevices()[0] == defaultRootDeviceInfo);
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:266]: I think there's a bit of a misunderstanding on what the default root device info does. It changes the IDeviceInfo attached to the default Client device. This lets you change things like its serial number, but it does not in any way affect its interactions with the module manager (scanning for devices, adding devices/function blocks...)

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 33]
```diff
@@ -0,0 +1,291 @@
...
+int main(int argc, const char* argv[])
+{
+    daq::InstanceBuilderPtr instanceBuilder = daq::InstanceBuilder().setRootDevice("daqref://device0");
+    daq::InstancePtr instance = instanceBuilder.build();
+
+    assert(instance.getAvailableDevices()[0] == instanceBuilder);
+
+    return 0;
+}
+----
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:291]: Connecting to a device is a separate how-to. This how-to should focus on creating and configuring the instance, not using it post-creation.

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]
Good work implementation-wise, the documentation still needs some tweaking.

File: core/opendaq/opendaq/include/opendaq/instance_builder.h (hunk 8 above)
Comment (denise-opendaq) [reply core/opendaq/opendaq/include/opendaq/instance_builder.h:246]: there is a comment in core\opendaq\opendaq\include\opendaq\instance_factory.h

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc [hunk 34]
```diff
@@ -207,28 +237,32 @@ Cpp::
...
+instance_builder = opendaq.InstanceBuilder()
+instance_builder.scheduler = scheduler
+instance = instance_builder.build()
+----
 ====
 
-== Configure openDAQ(TM) Root Device
-If, after creating the Instance object, the developer has not set the root device, the Instance will use the Client device as the root. Using the Instance Builder, developers can discover and connect to available devices using the method `getAvailableDevices`. In the Instance Builder, developers can forcibly set the available device.
+== Configure openDAQ(TM) Default Root Device
+The Instance has the client device as the default root device. A developer can modify the default device by setting the default root device info and local id in the instance builder.
```
Comment (JakaMohorkoDS) [inline docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:265]: I'd add here that my setting a new default root device via the connection string, on Instance build, the module manager will be used to create/connect to the device with the given connection string. Said device will then be set as the new root device.

**Review by JakaMohorkoDS — APPROVED** [review]
Lgtm, only added 1 extra remark.

File: docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc (hunk 34 above)
Comment (denise-opendaq) [reply docs/Antora/modules/howto_guides/pages/howto_configure_instance.adoc:265]: this is written in "Configure openDAQ(TM) Root Device"
